### PR TITLE
`/repository/files/{UUID}` should yield 404 for non-existing files (#355, #448)

### PR DESCRIPTION
**Pull request by GPelayo**

### CODE REVIEW

File: src/azul/service/repository.py [hunk 1]
```diff
@@ -1,12 +1,27 @@
...
 
 from azul.service import AbstractService
 from azul.service.responseobjects.elastic_request_builder import ElasticTransformDump as EsTd
 
 FileUrlFunc = Callable[[str, Mapping[str, Any]], str]
 
+
+class UUIDNotFoundError(Exception):
+    def __init__(self, entity_type: str, item_id: str):
+        Exception.__init__(self)
```
Comment (hannes-ucsc) [inline src/azul/service/repository.py]: This not how you call the overridden method in Python 3.6.

File: src/azul/service/repository.py [hunk 2]
```diff
@@ -1,12 +1,27 @@
...
 from typing import Callable, Mapping, Any
+import uuid
 
 from azul.service import AbstractService
 from azul.service.responseobjects.elastic_request_builder import ElasticTransformDump as EsTd
 
 FileUrlFunc = Callable[[str, Mapping[str, Any]], str]
 
+
+class UUIDNotFoundError(Exception):
```
Comment (hannes-ucsc) [inline src/azul/service/repository.py]:
```suggestion
class EntityNotFoundError(Exception):
```

File: src/azul/service/repository.py [hunk 3]
```diff
@@ -1,12 +1,27 @@
...
+    def __init__(self, entity_type: str, item_id: str):
```
Comment (hannes-ucsc) [inline src/azul/service/repository.py]:
```suggestion
    def __init__(self, entity_type: str, entity_id: str):
```

File: src/azul/service/repository.py [hunk 4]
```diff
@@ -28,10 +43,18 @@ def _get_data(self, entity_type, pagination, filters, file_url_func):
...
             filters['file']['projectId'] = {"is": [item_id]}
+        elif entity_type == 'specimens':
+            filters['file']['specimenDocumentId'] = {"is": [item_id]}
         else:
             filters['file']['fileId'] = {"is": [item_id]}
+
+        try:
+            if item_id != str(uuid.UUID(item_id)):
+                raise InvalidUUIDError(item_id)
+        except ValueError:
```
Comment (hannes-ucsc) [inline src/azul/service/repository.py]: Limit the body of the `try:` block such that it only contains the invocation of `uuid.UUID()`

File: src/azul/service/repository.py [hunk 5]
```diff
@@ -1,12 +1,27 @@
...
 from azul.service import AbstractService
 from azul.service.responseobjects.elastic_request_builder import ElasticTransformDump as EsTd
 
 FileUrlFunc = Callable[[str, Mapping[str, Any]], str]
 
+
+class UUIDNotFoundError(Exception):
+    def __init__(self, entity_type: str, item_id: str):
+        Exception.__init__(self)
+        self.message = f"Can't find an entity in {entity_type} with an uuid, {item_id}."
```
Comment (hannes-ucsc) [inline src/azul/service/repository.py]: You can pass the message to the overridden constructor directly.

File: test/service/test_request_validation.py [hunk 6]
```diff
@@ -153,6 +153,20 @@ def test_valid_sort_facet_but_bad_filter_facet_of_file(self):
         self.assertEqual(400, response.status_code, response.json())
         self.assertEqual(self.filter_facet_message, response.json())
 
+    def test_single_entity_error_responses(self):
+        self._index_canned_bundle(("f79257a7-dfc6-46d6-ae00-ba4b25313c10", "2018-09-14T133314.453337Z"))
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: Is this necessary? This test isn't actually hitting the index.

File: test/service/test_request_validation.py [hunk 7]
```diff
@@ -153,6 +153,20 @@ def test_valid_sort_facet_but_bad_filter_facet_of_file(self):
...
+        entities = ['specimens', 'files', 'projects']
+        expected_values = [('2b7959bb-acd1-4aa3-9557-345f9b3c6327', 404),
+                           ('-0c5ac7c0-817e-40d4-b1b1-34c3d5cfecdb-', 400),
+                           ('FOO', 400)]
+
+        for uuid, expected_error_code in expected_values:
+            for entity_name in entities:
+                with self.subTest(entity_name, entity_name=entity_name, error_code=expected_error_code, uuid=uuid):
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: Why is entity_name passed twice?

File: test/service/test_request_validation.py [hunk 8]
```diff
@@ -153,6 +153,20 @@ def test_valid_sort_facet_but_bad_filter_facet_of_file(self):
         self.assertEqual(400, response.status_code, response.json())
         self.assertEqual(self.filter_facet_message, response.json())
 
+    def test_single_entity_error_responses(self):
+        self._index_canned_bundle(("f79257a7-dfc6-46d6-ae00-ba4b25313c10", "2018-09-14T133314.453337Z"))
+        entities = ['specimens', 'files', 'projects']
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]:
```suggestion
        entity_types = ['specimens', 'files', 'projects']
```

File: test/service/test_request_validation.py [hunk 9]
```diff
@@ -153,6 +153,20 @@ def test_valid_sort_facet_but_bad_filter_facet_of_file(self):
...
+        expected_values = [('2b7959bb-acd1-4aa3-9557-345f9b3c6327', 404),
+                           ('-0c5ac7c0-817e-40d4-b1b1-34c3d5cfecdb-', 400),
+                           ('FOO', 400)]
+
+        for uuid, expected_error_code in expected_values:
+            for entity_name in entities:
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]:
```suggestion
            for entity_type in entity_types:
```

The degree of variation and inconsistency in your variable naming is impressive.

File: test/service/test_request_validation.py [hunk 10]
```diff
@@ -153,6 +153,20 @@ def test_valid_sort_facet_but_bad_filter_facet_of_file(self):
...
         self.assertEqual(self.filter_facet_message, response.json())
 
+    def test_single_entity_error_responses(self):
+        self._index_canned_bundle(("f79257a7-dfc6-46d6-ae00-ba4b25313c10", "2018-09-14T133314.453337Z"))
+        entities = ['specimens', 'files', 'projects']
+        expected_values = [('2b7959bb-acd1-4aa3-9557-345f9b3c6327', 404),
+                           ('-0c5ac7c0-817e-40d4-b1b1-34c3d5cfecdb-', 400),
+                           ('FOO', 400)]
+
+        for uuid, expected_error_code in expected_values:
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: Bonus points for this approach.

Comment (hannes-ucsc) [reply test/service/test_request_validation.py]: I would inline `expected_values`

File: src/azul/service/repository.py [hunk 11]
```diff
@@ -28,10 +43,18 @@ def _get_data(self, entity_type, pagination, filters, file_url_func):
     def _get_item(self, entity_type, item_id, pagination, filters, file_url_func):
         if entity_type == 'projects':
             filters['file']['projectId'] = {"is": [item_id]}
+        elif entity_type == 'specimens':
+            filters['file']['specimenDocumentId'] = {"is": [item_id]}
```
Comment (hannes-ucsc) [inline src/azul/service/repository.py]: These two lines should be in a separate commit. They fix a serious bug.

File: test/service/test_request_validation.py (hunk 7 above)
Comment (GPelayo) [reply test/service/test_request_validation.py]: That was a typo.

File: test/service/test_request_validation.py (hunk 6 above)
Comment (GPelayo) [reply test/service/test_request_validation.py]: I thought that I needed a valid bundle to test (`-0c5ac7c0-817e-40d4-b1b1-34c3d5cfecdb-', 400). But after thinking it over, I realized the canned bundle isn't necessary.

File: lambdas/service/app.py [hunk 12]
```diff
@@ -153,10 +153,10 @@ def repository_search(entity_type: str, item_id: str):
         pagination = _get_pagination(app.current_request, entity_type)
         service = RepositoryService()
         return service.get_data(entity_type, pagination, filters, item_id, file_url)
-    except BadArgumentException as e:
-        raise BadRequestError(msg=e.message)
-    except IndexNotFoundError as e:
-        raise NotFoundError(msg=e.message)
+    except (BadArgumentException, InvalidUUIDError) as e:
+        raise BadRequestError(msg=e)
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: This is relying on an implementation detail, specifically that BadRequestError's constructor implicitly applies `str()` to the argument.

```suggestion
        raise BadRequestError(msg=str(e))
```
